### PR TITLE
[JENKINS-58041] Add support for IPv6 when getting "Jenkins URL"

### PR DESCRIPTION
**Pull request by lorainsu**
See [JENKINS-58041](https://issues.jenkins-ci.org/browse/JENKINS-58041).

### Proposed changelog entries

* Issue:  Add support for IPv6 when getting "Jenkins URL"， since the default function "Jenkins.getRootUrlFromRequest()" do not support IPv6 address.

### Submitter checklist

- [x] JIRA issue is well described
- [x] Changelog entry appropriate for the audience affected by the change (users or developer, depending on the change). [Examples](https://github.com/jenkins-infra/jenkins.io/blob/master/content/_data/changelogs/weekly.yml)
      * Use the `Internal: ` prefix if the change has no user-visible impact (API, test frameworks, etc.)
- [x] Appropriate autotests or explanation to why this change has no tests
- [x] For dependency updates: links to external changelogs and, if possible, full diffs

### Desired reviewers

### CODE REVIEW

**Review by jvz — APPROVED** [review]

**Review by jvz — APPROVED** [review]

**Review by jvz — APPROVED** [review]

**Review by jvz — APPROVED** [review]

**Review by jvz — APPROVED** [review]

**Review by jvz — APPROVED** [review]

**Comment by jvz** [comment]
Damn github, sorry for spam.

**Comment by Wadeck** [comment]
Due to 
> [...] In fact I recommend deprecating getRootUrlFromRequest as a bad idea from the start [...]
From #4072.

I think this PR should require particular attention :)

**Review by jeffret-b — APPROVED** [review]
Looks like it should work fine.

There been comments about re-working or getting rid of getRootUrl(), but I don't see any reason this one couldn't proceed until something else is done there.

File: core/src/main/java/jenkins/model/Jenkins.java
Comment (oleg-nenashev) [inline core/src/main/java/jenkins/model/Jenkins.java]: any chance it changes the behavior for IPv4.
`buf.append(host, 0, index)` may start behaving differently if there are two semicolons in the hostname. Not sure it is possible.

**Review by oleg-nenashev — APPROVED** [review]
Just one question to clarify before merging

File: core/src/main/java/jenkins/model/Jenkins.java
Comment (jeffret-b) [reply core/src/main/java/jenkins/model/Jenkins.java]: I don't think it would change that behavior. Not all hostnames abide by the RFCs, but the scenario you describe doesn't seem possible or reasonable.

**Comment by oleg-nenashev** [comment]
Will merge it tomorrow if no negative feedback

**Comment by oleg-nenashev** [comment]
Thanks for your contribution @lorainsu !

**Pull request merged**
